aten/src/ATen/native/mkldnn/Matmul.cpp

```diff
@@ -0,0 +1,75 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/native/mkldnn/Matmul.h>
+#if !AT_MKLDNN_ENABLED()
+
+namespace at {
+namespace native {
+
+void mkldnn_matmul(
+    const Tensor &mat1,
+    const Tensor &mat2,
+    Tensor &result,
+    float beta,
+    float alpha) {
+  TORCH_CHECK(false, "mkldnn_matmul: ATen not compiled with MKLDNN support");
+}
+} // namespace native
+} // namespace at
+
+#else // AT_MKLDNN_EBABLED
+
+#include <ATen/native/mkldnn/MKLDNNCommon.h>
+#include <ATen/native/mkldnn/Utils.h>
+
+namespace at {
+namespace native {
+
+void mkldnn_matmul(
+    const Tensor &mat1,
+    const Tensor &mat2,
+    const Tensor &result,
+    float beta,
+    float alpha) {
+  TORCH_CHECK(mat1.dim() == 2 || mat1.dim() == 3, "mkldnn_matmul:  expect mat1 to be 2-D or 3-D tensor");
+  TORCH_CHECK(mat2.dim() == 2 || mat2.dim() == 3, "mkldnn_matmul:  expect mat2 to be 2-D or 3-D tensor");
+  TORCH_CHECK(mat1.scalar_type() == at::kBFloat16, "mkldnn_matmul:  only enabled for bf16 path");
+  TORCH_CHECK(mat2.scalar_type() == at::kBFloat16, "mkldnn_matmul:  only enabled for bf16 path");
+  TORCH_CHECK(result.scalar_type() == at::kBFloat16, "mkldnn_matmul:  only enabled for bf16 path");
+  TORCH_CHECK(mkldnn_bf16_device_check(),
+    "mkldnn_matmul: mkldnn_matmul bf16 path needs the cpu support avx512bw, avx512vl and avx512dq");
+  // In aten, expect beta * result = 0 if beta = 0
+  ideep::attr_t op_attr;
+  if (beta != 0.0f) op_attr = ideep::attr_t::fuse_sum();
+  // If alpha = 0, dose not need actually do gemm computation
+  if (alpha == 0)
+    return;
+  Tensor mat1_ = mat1.unsafeGetTensorImpl()->is_wrapped_number() ? mat1.contiguous() : mat1;
+  Tensor mat2_ = mat1.unsafeGetTensorImpl()->is_wrapped_number() ? mat2.contiguous() : mat2;
```

Comment:
first i don't quite understand why we need to call `unsafeGetTensorImpl()->is_wrapped_number()` here.
second `mat.contiguous()` will create additional copy for the case when the tensor is 'transposed' contiguous, any problem if we directly create ideep::tensor based on size and stride for this case?

Reply:
It is because mkldnn dose not optimize these formats and will run into "ref" path.